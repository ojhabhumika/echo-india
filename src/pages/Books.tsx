import React, { useState, useEffect } from "react"
import axios from "axios"
import { PageHeader, Table, Input } from "antd"
import { useLocation } from "react-router-dom"
const { Search } = Input

interface book {
  id: number
  book_author: string
  book_title: string
  book_publication_year: number
  book_publication_country: string
  book_publication_city: string
  book_pages: number
}

interface getBooksPayload {
  page: number
  itemsPerPage: number
  filters?: any[]
}

const Books: React.FC = () => {

  const defaultPage: number = 1
  const defaultPageSize: number = 10

  let pageQuery: number = +(useQuery("page")) || defaultPage
  let pageSizeQuery: number = +(useQuery("pageSize")) || defaultPageSize

  const [loading, setLoading] = useState<boolean>(false)
  const [books, setBooks] = useState<book[]>([])
  const [page, setPage] = useState<number>(pageQuery)
  const [pageSize, setPageSize] = useState<number>(pageSizeQuery)
  const [totalBooksCount, setTotalBooksCount] = useState<number>(0)
  const [filterText, setFilterText] = useState<string>("")

  const columns = [
    {
      title: "#",
      render: (value: any, item: any, index: number) => (page - 1) * pageSize + index + 1,
    },
    {
      title: "Title",
      dataIndex: "book_title",
    },
    {
      title: "Author",
      dataIndex: "book_author",
      render: (text: string[]) => text[0],
    },
    {
      title: "Page Count",
      dataIndex: "book_pages",
    },
    {
      title: "Publication City",
      dataIndex: "book_publication_city",
    },
    {
      title: "Publication Country",
      dataIndex: "book_publication_country",
    },
    {
      title: "Publication Year",
      dataIndex: "book_publication_year",
    },
  ]

  useEffect(() => {
    getBooksData()
  }, [page, pageSize])

  useEffect(() => {
    (page > 1 && filterText) ? setPage(1) : getBooksData()
  }, [filterText])

  const getBooksData = async () => {
    setLoading(true)
    setBooks([])
    try {

      window.history.replaceState(null, "", `?page=${page}&pageSize=${pageSize}`)

      const payload: getBooksPayload = {
        page,
        itemsPerPage: pageSize,
      }

      if (filterText) {
        payload.filters = [{ type: "all", values: [filterText] }]
      }

      const res: any = await axios.post(`http://nyx.vima.ekt.gr:3000/api/books`, payload)
      const { books, count } = res.data
      //console.table(books[0])
      setBooks(books)
      setTotalBooksCount(count)
      window.scrollTo(0, 0)

    } catch (err) {
      console.log("err :>> ", err)
    } finally {
      setLoading(false)
    }
  }

  return (
    <div style={{ margin: "10px 35px" }} data-testid="test-books" >
      <PageHeader
        onBack={() => alert("In Progress")}
        title="Books"
        subTitle="Search the world's most comprehensive index of full-text books."
        extra={[
          <Search
            key="1"
            placeholder="Search Books.."
            onSearch={(value: string) => setFilterText(value)}
            enterButton
            allowClear
            size="large"
          />,
        ]}
      />

      <Table
        rowKey="id"
        columns={columns}
        loading={loading}
        dataSource={books}
        pagination={{
          current: page,
          pageSize,
          total: totalBooksCount,
          onChange: (page) => setPage(page),
          onShowSizeChange: (current, pageSize) => setPageSize(pageSize),
        }}
      />
    </div>
  )
}

const useQuery = (name: string): string => {
  const queryValue = new URLSearchParams(useLocation().search).get(name)
  return queryValue ?? ""
}

export default Books



