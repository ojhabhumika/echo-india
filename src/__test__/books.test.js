import { render, waitFor, cleanup } from '@testing-library/react'
import Books from '../pages/Books'
import axios from 'axios'
window.scrollTo = jest.fn();

jest.mock('react-router-dom', () => ({
    useLocation: () => new URLSearchParams({
        pathname: '/books',
        search: '?page=1'
    })
}))

jest.mock('axios')

Object.defineProperty(window, 'matchMedia', {
    value: () => {
        return {
            matches: false,
            addListener: () => { },
            removeListener: () => { }
        }
    }
})

afterEach(cleanup)

test("should render Books component", async () => {

    let response = {
        count: 2,
        books: [{
            book_author: ['Ανώνυμος'],
            book_pages: 104,
            book_publication_city: "Βενετία",
            book_publication_country: "Ιταλία",
            book_publication_year: 1529,
            book_title: "Ο Αλέξανδρος ο Μακεδών",
            id: 2086
        },
        {
            book_author: ['Ανώνυμος'],
            book_pages: 104,
            book_publication_city: "Βενετία",
            book_publication_country: "Ιταλία",
            book_publication_year: 1529,
            book_title: "Ο Αλέξανδρος ο Μακεδών",
            id: 2087
        }]
    }

    axios.post.mockResolvedValue({
        data: response,
    })

    const { getByTestId, asFragment } = render(<Books />);

    await waitFor(() => {
        const listNode = getByTestId('test-books')
        expect(listNode).toBeDefined();
        expect(asFragment(<Books />)).toMatchSnapshot()
    })

})