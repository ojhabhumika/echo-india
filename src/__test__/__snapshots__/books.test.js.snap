// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Books component 1`] = `
<DocumentFragment>
  <div
    data-testid="test-books"
    style="margin: 10px 35px;"
  >
    <div
      class="ant-page-header ant-page-header-ghost"
    >
      <div
        class="ant-page-header-heading"
      >
        <div
          class="ant-page-header-heading-left"
        >
          <div
            class="ant-page-header-back"
          >
            <div
              aria-label="Back"
              class="ant-page-header-back-button"
              role="button"
              style="border: 0px; background: transparent; padding: 0px; line-height: inherit; display: inline-block;"
              tabindex="0"
            >
              <span
                aria-label="arrow-left"
                class="anticon anticon-arrow-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="arrow-left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 000 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <span
            class="ant-page-header-heading-title"
            title="Books"
          >
            Books
          </span>
          <span
            class="ant-page-header-heading-sub-title"
            title="Search the world's most comprehensive index of full-text books."
          >
            Search the world's most comprehensive index of full-text books.
          </span>
        </div>
        <span
          class="ant-page-header-heading-extra"
        >
          <span
            class="ant-input-group-wrapper ant-input-group-wrapper-lg ant-input-search ant-input-search-large ant-input-search-with-button"
          >
            <span
              class="ant-input-wrapper ant-input-group"
            >
              <span
                class="ant-input-affix-wrapper ant-input-affix-wrapper-lg"
              >
                <input
                  class="ant-input ant-input-lg"
                  placeholder="Search Books.."
                  type="text"
                  value=""
                />
                <span
                  class="ant-input-suffix"
                >
                  <span
                    aria-label="close-circle"
                    class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                    role="button"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close-circle"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                      />
                    </svg>
                  </span>
                </span>
              </span>
              <span
                class="ant-input-group-addon"
              >
                <button
                  class="ant-btn ant-btn-primary ant-btn-lg ant-input-search-button"
                  type="button"
                >
                  <span
                    aria-label="search"
                    class="anticon anticon-search"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="search"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                      />
                    </svg>
                  </span>
                </button>
              </span>
            </span>
          </span>
        </span>
      </div>
    </div>
    <div
      class="ant-table-wrapper"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div>
          <div
            class="ant-spin ant-spin-spinning"
          >
            <span
              class="ant-spin-dot ant-spin-dot-spin"
            >
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
            </span>
          </div>
        </div>
        <div
          class="ant-spin-container ant-spin-blur"
        >
          <div
            class="ant-table ant-table-empty"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: auto;"
                >
                  <colgroup />
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell"
                      >
                        #
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Title
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Author
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Page Count
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Publication City
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Publication Country
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Publication Year
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-placeholder"
                    >
                      <td
                        class="ant-table-cell"
                        colspan="7"
                      >
                        <div
                          class="ant-empty ant-empty-normal"
                        >
                          <div
                            class="ant-empty-image"
                          >
                            <svg
                              class="ant-empty-img-simple"
                              height="41"
                              viewBox="0 0 64 41"
                              width="64"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <g
                                fill="none"
                                fill-rule="evenodd"
                                transform="translate(0 1)"
                              >
                                <ellipse
                                  class="ant-empty-img-simple-ellipse"
                                  cx="32"
                                  cy="33"
                                  rx="32"
                                  ry="7"
                                />
                                <g
                                  class="ant-empty-img-simple-g"
                                  fill-rule="nonzero"
                                >
                                  <path
                                    d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                  />
                                  <path
                                    class="ant-empty-img-simple-path"
                                    d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  />
                                </g>
                              </g>
                            </svg>
                          </div>
                          <div
                            class="ant-empty-description"
                          >
                            No Data
                          </div>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
